### PR TITLE
Generalize view and size functions

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Generalizes inputs to view and size functions (`col`, `row`, `sub_col`, `sub_row`, `head`, `tail`, `diagonal`, `transpose`, `cols`, `rows`, `num_elements`, `dims`).

## Tests
This is a refactor so no new tests.

## Side Effects
None.

## Checklist

- [x] Math issue #1470 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by t4c1** [comment]
I think this error will be gone once #1628 is merged in.

**Comment by t4c1** [comment]
This is now ready for a review.

File: stan/math/prim/err/check_vector_index.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * by setting <code>stan::error_index::value</code>.
+ * @tparam T Vector type
+ * @param function Function name (for error messages)
+ * @param name Variable name (for error messages)
+ * @param y vector to test
+ * @param i row index to check
+ * @throw <code>std::out_of_range</code> if the index is out of range.
+ */
+template <typename T, typename = require_eigen_vector_t<T>>
+inline void check_vector_index(const char* function, const char* name,
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_vector_index.hpp]: This function is missing tests.

File: stan/math/prim/fun/col.hpp [hunk 2]
```diff
@@ -14,17 +14,16 @@ namespace math {
...
  * @throw std::out_of_range if j is out of range.
  */
-template <typename T>
-inline Eigen::Matrix<T, Eigen::Dynamic, 1> col(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m, size_t j) {
+template <typename T, typename = require_eigen_t<T>>
+inline auto col(const T& m, size_t j) {
   check_column_index("col", "j", m, j);
-  return m.col(j - 1);
+  return m.col(j - 1).eval();
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/col.hpp]: Please add a "xyz_works_with_other_functions" test for this function so someone doesnt remove the eval for some reason before we start accepting eigen expressions.

File: stan/math/prim/fun/diagonal.hpp [hunk 3]
```diff
@@ -10,14 +11,13 @@ namespace math {
...
  * @param m Specified matrix.
  * @return Diagonal of the matrix.
  */
-template <typename T>
-inline Eigen::Matrix<T, Eigen::Dynamic, 1> diagonal(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m) {
-  return m.diagonal();
+template <typename T, typename = require_eigen_t<T>>
+inline auto diagonal(const T& m) {
+  return m.diagonal().eval();
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/diagonal.hpp]: Please add a "xyz_works_with_other_functions" test

File: stan/math/prim/fun/head.hpp [hunk 4]
```diff
@@ -9,41 +9,21 @@ namespace stan {
...
- * @throw std::out_of_range if n is out of range.
- */
-template <typename T>
-inline Eigen::Matrix<T, 1, Eigen::Dynamic> head(
-    const Eigen::Matrix<T, 1, Eigen::Dynamic>& rv, size_t n) {
-  if (n != 0) {
-    check_column_index("head", "n", rv, n);
-  }
-  return rv.head(n);
+  return v.head(n).eval();
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/head.hpp]: Please add a "xyz_works_with_other_functions" test

File: stan/math/prim/fun/row.hpp [hunk 5]
```diff
@@ -14,18 +14,17 @@ namespace math {
...
  */
-template <typename T>
-inline Eigen::Matrix<T, 1, Eigen::Dynamic> row(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m, size_t i) {
+template <typename T, typename = require_eigen_t<T>>
+inline auto row(const T& m, size_t i) {
   check_row_index("row", "i", m, i);
 
-  return m.row(i - 1);
+  return m.row(i - 1).eval();
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/row.hpp]: Please add a "xyz_works_with_other_functions" test

File: stan/math/prim/fun/sub_col.hpp [hunk 6]
```diff
@@ -10,23 +10,21 @@ namespace math {
...
-    size_t j, size_t nrows) {
+template <typename T, typename = require_eigen_t<T>>
+inline auto sub_col(const T& m, size_t i, size_t j, size_t nrows) {
   check_row_index("sub_col", "i", m, i);
   if (nrows > 0) {
     check_row_index("sub_col", "i+nrows-1", m, i + nrows - 1);
   }
   check_column_index("sub_col", "j", m, j);
-  return m.block(i - 1, j - 1, nrows, 1);
+  return m.col(j - 1).segment(i - 1, nrows).eval();
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/sub_col.hpp]: Please add a "xyz_works_with_other_functions" test

File: stan/math/prim/fun/sub_row.hpp [hunk 7]
```diff
@@ -10,23 +10,21 @@ namespace math {
...
-    size_t j, size_t ncols) {
+template <typename T, typename = require_eigen_t<T>>
+inline auto sub_row(const T& m, size_t i, size_t j, size_t ncols) {
   check_row_index("sub_row", "i", m, i);
   check_column_index("sub_row", "j", m, j);
   if (ncols > 0) {
     check_column_index("sub_col", "j+ncols-1", m, j + ncols - 1);
   }
-  return m.block(i - 1, j - 1, 1, ncols);
+  return m.row(i - 1).segment(j - 1, ncols).eval();
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/sub_row.hpp]: Please add a "xyz_works_with_other_functions" test

File: stan/math/prim/fun/tail.hpp [hunk 8]
```diff
@@ -10,41 +10,21 @@ namespace stan {
...
-template <typename T>
-inline Eigen::Matrix<T, 1, Eigen::Dynamic> tail(
-    const Eigen::Matrix<T, 1, Eigen::Dynamic>& rv, size_t n) {
+inline auto tail(const T& v, size_t n) {
   if (n != 0) {
-    check_column_index("tail", "n", rv, n);
+    check_vector_index("tail", "n", v, n);
   }
-  return rv.tail(n);
+  return v.tail(n).eval();
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/tail.hpp]: Please add a "xyz_works_with_other_functions" test

File: stan/math/prim/fun/col.hpp (hunk 2 above)
Comment (t4c1) [reply stan/math/prim/fun/col.hpp]: Which other function should I test this with? In #1644 I used multiply, but it has been generalized in #1628, so these tests won't fail anymore if eval is removed. The problem is that regardless of what function we choose such tests won't detect missing eval anymore once the function is generalized. I could make one non-generalized function just for testing, but then we will have to remove many tests once all functions accept expressions.

Comment (rok-cesnovar) [reply stan/math/prim/fun/col.hpp]: Yeah, that is a fair point. Whichever function we use to test with is going to not be a valuable test once its generalized. Well I guess there isnt a good solution here.

Comment (rok-cesnovar) [reply stan/math/prim/fun/col.hpp]: And we need to test this on the upstream levels, which we said before. Will resolve these then.

**Pull request merged**